src/main/java/com/laytonsmith/core/constructs/InstanceofUtil.java

```diff
@@ -50,131 +72,329 @@
+	 * <li>Nothing is instanceof void and null.</li>
+	 * <li>void is instanceof nothing.</li>
+	 * <li>{@code A<B>} is instanceof {@code A}, because {@code A} is {@code A<auto>}.</li>
 	 * </ul>
-	 * @param type - The type to check for.
-	 * Java {@code null} can be used to indicate no type (e.g. from control flow breaking statements).
-	 * @param instanceofThis - The {@link CClassType} to check against.
-	 * Java {@code null} can be used to indicate that anything is allowed to match this
-	 * (i.e. making this method return {@code true}).
+	 *
+	 * @param type - The type to check for. Java {@code null} can be used to indicate no type (e.g. from control flow
+	 * breaking statements).
+	 * @param instanceofThis - The {@link CClassType} to check against. Java {@code null} can be used to indicate that
+	 * anything is allowed to match this (i.e. making this method return {@code true}).
+	 * @param instanceofThisGenerics The LHS generics. Confusingly, this is actually on the RHS of the instanceof
+	 * statement, because we generally accept LHS statements RHS of the instanceof. For example
+	 * {@code (new A<int>()) instanceof A<? extends primitive>} and {@code (new A<int>()) instanceof A<int>} are both
+	 * valid. In the second example, this is simply a LeftHandGenericUse with an ExactType value.
 	 * @param env
-	 * @return
+	 * @return {@code true} if type is instance of instanceofThis.
 	 */
-	public static boolean isInstanceof(CClassType type, CClassType instanceofThis, Environment env) {
+	public static boolean isInstanceof(
+			CClassType type, CClassType instanceofThis, LeftHandGenericUse instanceofThisGenerics, Environment env) {
+//		instanceofThis = (instanceofThis != null ? CClassType.getNakedClassType(instanceofThis.getFQCN(), env) : null);
 
 		// Handle special cases.
-		if(type == instanceofThis || instanceofThis == null || type == CClassType.AUTO
-				|| type == CNull.TYPE || instanceofThis == CClassType.AUTO) {
+		if((type == instanceofThis && instanceofThisGenerics == null) // Identity short circuit
+				|| instanceofThis == null // java null on RHS defined as true for implementation purposes
+				|| (instanceofThis.equals(type) && instanceofThis.getGenericDeclaration() == null) // no generics involved, and the types are equal
+				|| CClassType.AUTO.equals(type) // auto type on
+				|| CClassType.AUTO.equals(instanceofThis) // either side
+				) {
 			return true;
 		}
-		if(type == null || type == CVoid.TYPE || instanceofThis == CVoid.TYPE || instanceofThis == CNull.TYPE) {
+		if(type == null // type is java null defined as false (except if instanceofThis was true, which is caught above)
+				|| CVoid.TYPE.equals(type) // void is not instanceof anything
+				|| CVoid.TYPE.equals(instanceofThis) // nothing is instanceof void
+				|| CNull.TYPE.equals(instanceofThis) // nothing is instanceof null (should be compile error)
+				|| CNull.TYPE.equals(type) // type is mscript null defined as false
```

Comment:
Ah, now I see why you've added the null check in `StaticAnalysis`. This method used to only be used by SA, and all instanceof checks there check whether a type is usable as some parameter/variable type, in which case `CNull` is suitable. But this is a fair change, as it actually isn't instance of anything. If you've added null checks in all places where SA uses this method, then this is all good.

Reply:
There is still work to be done here, if you notice, this and the method below are nearly identical. They need to be merged and harmonized into a single implementation.